homeassistant/components/refoss/coordinator.py

```diff
@@ -0,0 +1,279 @@
+"""Coordinators for the Refoss integration."""
+from typing import TypeVar, List, Optional, Iterable
+import async_timeout
+import threading
+import asyncio
+from asyncio import AbstractEventLoop
+from datetime import timedelta
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.helpers.typing import HomeAssistantType
+
+from refoss_ha.socket_util import pushStateDataList
+from refoss_ha.http_device import HttpDeviceInfo
+from refoss_ha.controller.device import BaseDevice
+from refoss_ha.socket_util import SocketUtil
+from refoss_ha.enums import Namespace
+from refoss_ha.controller.toggle import ToggleXMix
+from refoss_ha.controller.system import SystemAllMixin
+from refoss_ha.const import LOGGER, PUSH, DOMAIN
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+
+T = TypeVar("T", bound=BaseDevice)
+
+
+_ABILITY_MATRIX = {
+    Namespace.CONTROL_TOGGLEX.value: ToggleXMix,
+    Namespace.SYSTEM_ALL.value: SystemAllMixin,
+}
+
+
+class RefossCoordinator(DataUpdateCoordinator):
+    def __init__(
+        self,
+        hass: HomeAssistantType,
+        config_entry: ConfigEntry,
+        update_interval: timedelta,
+        loop: Optional[AbstractEventLoop] = None,
+    ):
+        self._entry = config_entry
+        self._devices_by_internal_id = {}
+        self._setup_done = False
+        self.socket = SocketUtil()
+        self._loop = asyncio.get_event_loop() if loop is None else loop
+
+        super().__init__(
+            hass=hass,
+            logger=LOGGER,
+            name="refoss_coordinator",
+            update_interval=update_interval,
+            update_method=self._async_fetch_data,
+        )
+
+    async def initial_setup(self):
+        if self._setup_done:
+            raise ValueError("This coordinator was already set up")
+
+        # Listening for socket messages
+        self.socket.startReveiveMsg()
+
+        t = threading.Thread(target=self.HandlePushState, args=(self.socket.event,))
+        t.start()
```

Comment:
Ideally this is asyncio instead.  We generally don't want to start new threads as this is a legacy design which doesn't scale well when a user has many devices.

Reply:
aioshelly might be a good lib to look at on how to do this

https://github.com/home-assistant-libs/aioshelly

Reply:
Changed to  asyncio.create_task , Is it feasible?   @bdraco

Reply:
resolve